bin/src/_main.rs

```diff
@@ -139,110 +142,170 @@ pub async fn _main(
+                        if let Err(e) = middleware_k8s_client {
                             let message = format!(
-                                "The agent could not access k8s api after several attempts: {}",
+                                "unable to refresh client for k8s metadata watcher: {:?}",
                                 e
                             );
                             error!("{}", message);
                             panic!("{}", message);
                         }
-                    };
-                }
-
-                let pod_name = std::env::var("POD_NAME").ok();
-                let namespace = std::env::var("NAMESPACE").ok();
-                let pod_label = std::env::var("POD_APP_LABEL").ok();
-
-                let k8s_event_stream = match config.log.log_k8s_events {
-                    K8sTrackingConf::Never => None,
-                    K8sTrackingConf::Always => {
-                        let k8s_event_stream_feature_meta = set_up_leader(
-                            &namespace,
-                            &k8s_client,
-                            &pod_name,
-                            K8S_EVENTS_LEASE_NAME.to_string(),
-                        )
-                        .await;
-
-                        Some(K8sEventStream::new(
-                            k8s_client.clone(),
-                            pod_name.unwrap_or_default(),
-                            namespace.unwrap_or_default(),
-                            pod_label.unwrap_or_default(),
-                            Arc::new(k8s_event_stream_feature_meta),
-                        ))
-                    }
-                };
-
-                let pod_name = std::env::var("POD_NAME").ok();
-                let namespace = std::env::var("NAMESPACE").ok();
 
-                let metric_stats_stream = match config.log.log_metric_server_stats {
-                    K8sTrackingConf::Never => None,
-                    K8sTrackingConf::Always => {
-                        let created_metrics_stream_feature_meta = set_up_leader(
-                            &namespace,
-                            &k8s_client,
-                            &pod_name,
-                            REPORTER_LEASE_NAME.to_string(),
-                        )
-                        .await;
+                        let (stream_handle, abort_registration) = AbortHandle::new_pair();
+                        let (thread_handle, swap_intent) = match middleware
+                            .kick_over(middleware_k8s_client.unwrap(), node_name.as_deref())
+                        {
+                            Ok((stream, swap_intent)) => {
+                                debug!("k8s metadata watcher kicked over!");
+                                let abortable_stream = Abortable::new(stream, abort_registration);
+                                (
+                                    tokio::spawn(async move {
```

Comment:
I do not feel good about this additional spawn inside loop.  how about using a channel instead of abortable stream?

Reply:
outdated